File: Managed/NextTurn.UE.Runtime/Core/Log.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using System.Diagnostics.CodeAnalysis;
using System.Runtime.InteropServices;
using NextTurn.UE.Annotations;

namespace Unreal
{
    public static class Log
    {
        /// <summary>
        /// Writes an informational message to the console and log file.
        /// </summary>
        /// <param name="message">
        /// The message to write.
        /// </param>
        public static void Display(string message) => NativeMethods.Display(message);

        /// <summary>
        /// Writes an error message to the console and log file.
        /// </summary>
        /// <param name="message">
        /// The message to write.
        /// </param>
        public static void Error(string message) => NativeMethods.Error(message);

        /// <summary>
        /// Writes a fatal message to the console and log file and crashes.
        /// </summary>
        /// <param name="message">
        /// The message to write.
        /// </param>
        [DoesNotReturn]
        public static void Fatal(string message) => NativeMethods.Fatal(message);

        /// <summary>
        /// Writes an informational message to the log file.
        /// </summary>
        /// <param name="message">
        /// The message to write.
        /// </param>
        public static void Information(string message) => NativeMethods.Information(message);

        /// <summary>
        /// Writes a verbose message to the log file.
        /// </summary>
        /// <param name="message">
        /// The message to write.
        /// </param>
        public static void Verbose(string message) => NativeMethods.Verbose(message);

        /// <summary>
        /// Writes a very verbose message to the log file.
        /// </summary>
        /// <param name="message">
        /// The message to write.
        /// </param>
        public static void VeryVerbose(string message) => NativeMethods.VeryVerbose(message);

        /// <summary>
        /// Writes a warning message to the console and log file.
        /// </summary>
        /// <param name="message">
        /// The message to write.
        /// </param>
        public static void Warning(string message) => NativeMethods.Warning(message);

        [StructLayout(LayoutKind.Auto, CharSet = CharSet.Unicode)]
        private static class NativeMethods
        {
            [Calli]
            public static extern void Display(string message);

            [Calli]
            public static extern void Error(string message);

            [Calli]
            public static extern void Fatal(string message);

            [Calli]
            public static extern void Information(string message);

            [Calli]
            public static extern void Verbose(string message);

            [Calli]
            public static extern void VeryVerbose(string message);

            [Calli]
            public static extern void Warning(string message);
        }
    }
}

File: Source/NextTurnRuntime/Private/Core/Log.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

ENABLE_WARNINGS;

RESTORE_WARNINGS;
DEFINE_LOG_CATEGORY_STATIC(LogNextTurn, VeryVerbose, All);
ENABLE_WARNINGS;

#define TYPE_EXPORT_MANAGED_TYPE_NAME Log

TYPE_EXPORT_START

void EXPORT_CALL_CONV Display(const char16_t* Message)
{
	RESTORE_WARNINGS;
	UE_LOG(LogNextTurn, Display, TEXT("%s"), CHAR16_TO_TCHAR(Message));
	ENABLE_WARNINGS;
}

void EXPORT_CALL_CONV Error(const char16_t* Message)
{
	RESTORE_WARNINGS;
	UE_LOG(LogNextTurn, Error, TEXT("%s"), CHAR16_TO_TCHAR(Message));
	ENABLE_WARNINGS;
}

void EXPORT_CALL_CONV Fatal(const char16_t* Message)
{
	RESTORE_WARNINGS;
	UE_LOG(LogNextTurn, Fatal, TEXT("%s"), CHAR16_TO_TCHAR(Message));
	ENABLE_WARNINGS;
}

void EXPORT_CALL_CONV Information(const char16_t* Message)
{
	RESTORE_WARNINGS;
	UE_LOG(LogNextTurn, Log, TEXT("%s"), CHAR16_TO_TCHAR(Message));
	ENABLE_WARNINGS;
}

void EXPORT_CALL_CONV Verbose(const char16_t* Message)
{
	RESTORE_WARNINGS;
	UE_LOG(LogNextTurn, Verbose, TEXT("%s"), CHAR16_TO_TCHAR(Message));
	ENABLE_WARNINGS;
}

void EXPORT_CALL_CONV VeryVerbose(const char16_t* Message)
{
	RESTORE_WARNINGS;
	UE_LOG(LogNextTurn, VeryVerbose, TEXT("%s"), CHAR16_TO_TCHAR(Message));
	ENABLE_WARNINGS;
}

void EXPORT_CALL_CONV Warning(const char16_t* Message)
{
	RESTORE_WARNINGS;
	UE_LOG(LogNextTurn, Warning, TEXT("%s"), CHAR16_TO_TCHAR(Message));
	ENABLE_WARNINGS;
}

const FMemberSymbol Members[] =
{
	EXPORT_METHOD(Display),
	EXPORT_METHOD(Error),
	EXPORT_METHOD(Fatal),
	EXPORT_METHOD(Information),
	EXPORT_METHOD(Verbose),
	EXPORT_METHOD(VeryVerbose),
	EXPORT_METHOD(Warning),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;
